File: static/detector/region.js
/* Locate the glasses REGION for the CNN classifier. Port of pipeline/region.py.
 * Worn (face found): face box + eye/eyeglasses cascades fix the eye line, region
 * spans the face width across that band. Held/studio: padded segment() bbox.
 * Fallback: central 90%. Eye cascades are loaded once via loadEyes(). */
(function (root, factory) {
  var mod = factory(root);
  if (typeof module !== 'undefined' && module.exports) { module.exports = mod; }
  if (root) { root.DET = root.DET || {}; root.DET.region = mod; }
})(typeof self !== 'undefined' ? self : (typeof window !== 'undefined' ? window : null),
function (root) {
  'use strict';

  var _eye = [];        // [CascadeClassifier] eyeglasses, eye
  var _ready = false;

  function ready() { return _ready; }

  /** Load eye/eyeglasses cascades. bytes: {eyeglasses:Uint8Array, eye:Uint8Array} */
  function loadEyes(cv, bytes) {
    function mk(name, data) {
      try { cv.FS_unlink(name); } catch (e) { /* first time */ }
      cv.FS_createDataFile('/', name, data, true, false, false);
      var c = new cv.CascadeClassifier();
      c.load(name);
      return c;
    }
    _eye = [mk('eyeglasses.xml', bytes.eyeglasses), mk('eye.xml', bytes.eye)];
    _ready = true;
  }

  /** Median eye-centre y from cascades in the face's upper 62%, else a typical line. */
  function eyeLine(gray, face, cv) {
    var fallback = face.y + 0.38 * face.h;
    if (!_ready) { return fallback; }
    var rx = Math.max(0, face.x), ry = Math.max(0, face.y);
    var rw = Math.min(gray.cols - rx, face.w);
    var rh = Math.min(face.y + Math.floor(0.62 * face.h), gray.rows) - ry;
    if (rw <= 0 || rh <= 0) { return fallback; }
    var roi = gray.roi(new cv.Rect(rx, ry, rw, rh));
    var side = Math.max(12, Math.floor(0.12 * face.w));
    var ms = new cv.Size(side, side), empty = new cv.Size();
    var cys = [], i, j;
    for (i = 0; i < _eye.length; i += 1) {
      var rv = new cv.RectVector();
      _eye[i].detectMultiScale(roi, rv, 1.1, 4, 0, ms, empty);
      for (j = 0; j < rv.size(); j += 1) {
        var r = rv.get(j);
        cys.push(ry + r.y + r.height / 2);
      }
      rv.delete();
    }
    roi.delete();
    if (!cys.length) { return fallback; }
    cys.sort(function (a, b) { return a - b; });
    return cys[Math.floor(cys.length / 2)];
  }

  function clip(x0, y0, x1, y1, W, H) {
    x0 = Math.max(0, Math.round(x0)); y0 = Math.max(0, Math.round(y0));
    x1 = Math.min(W, Math.round(x1)); y1 = Math.min(H, Math.round(y1));
    return { x: x0, y: y0, w: Math.max(0, x1 - x0), h: Math.max(0, y1 - y0) };
  }

  /** @returns {{x,y,w,h,method}} region bbox in canonical pixels. */
  function locateRegion(pre, cfg, cv, face) {
    var H = pre.gray.rows, W = pre.gray.cols;
    var fd = root.DET && root.DET.facedet;
    if (face === undefined) {
      face = (fd && fd.ready && fd.ready()) ? fd.detectFace(pre.gray, cv) : null;
    }
    if (face) {
      var eyeCy = eyeLine(pre.gray, face, cv);
      var b = clip(face.x - cfg.region_wpad * face.w, eyeCy - cfg.region_up * face.h,
                   face.x + face.w + cfg.region_wpad * face.w, eyeCy + cfg.region_down * face.h, W, H);
      if (b.w > 20 && b.h > 20) { b.method = 'worn'; return b; }
    }
    var seg = root.DET.segment.segment(pre.grayEq, cfg, cv);
    var res;
    if (seg.valid && seg.method !== 'center-crop') {
      var bx = seg.bbox[0], by = seg.bbox[1], bw = seg.bbox[2], bh = seg.bbox[3];
      res = clip(bx - cfg.region_seg_pad * bw, by - cfg.region_seg_pad * bh,
                 bx + bw + cfg.region_seg_pad * bw, by + bh + cfg.region_seg_pad * bh, W, H);
      res.method = 'held';
    } else {
      res = clip(0.05 * W, 0.05 * H, 0.95 * W, 0.95 * H, W, H);
      res.method = 'center';
    }
    if (seg.free) { seg.free(); }
    return res;
  }

  return { loadEyes: loadEyes, ready: ready, locateRegion: locateRegion,
           eyeLine: eyeLine };
});

File: pipeline/region.py
"""Locate the glasses REGION to feed the CNN classifier.

Two localization paths (whichever fits), producing ONE bbox around the whole
glasses so the CNN sees the frame + both end pieces (where a Ray-Ban camera sits):

  * WORN path (face found): a Haar face box + eye/eyeglasses cascades fix the
    eye-line; the region spans the face width across that band.
  * HELD/STUDIO path (no face): the segment() frame bbox (dark frame on a
    lighter background), padded out a little.
  * Fallback: the central 90% crop.

The crop is later resized (with mild aspect distortion) to the CNN input size,
identically at train and inference, so no letterboxing parity to maintain.
Shared by train_region_clf.py (Python) and mirrored by static/detector/region.js.
"""
from __future__ import annotations

import os
from dataclasses import dataclass

import cv2
import numpy as np

from config import Config
from pipeline import facedet

_HAAR_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
                         "models", "haar")
_EYE_CACHE: dict = {}


def _eye_cascade(name: str):
    if name not in _EYE_CACHE:
        path = os.path.join(_HAAR_DIR, name)
        if not os.path.exists(path):
            path = cv2.data.haarcascades + name
        _EYE_CACHE[name] = cv2.CascadeClassifier(path)
    return _EYE_CACHE[name]


@dataclass
class Region:
    x: int
    y: int
    w: int
    h: int
    method: str          # "worn" | "held" | "center"


def _eye_line(gray: np.ndarray, face: facedet.Face):
    """Median eye-centre y from eye/eyeglasses cascades in the face's upper 62%,
    or a typical eye-line if none fire."""
    x, y, w, h = face.x, face.y, face.w, face.h
    y1 = y + int(0.62 * h)
    roi = gray[max(0, y):y1, max(0, x):x + w]
    if roi.size == 0:
        return y + 0.38 * h
    ms = (max(12, int(0.12 * w)),) * 2
    cys = []
    for name in ("haarcascade_eye_tree_eyeglasses.xml", "haarcascade_eye.xml"):
        for (ex, ey, ew, eh) in _eye_cascade(name).detectMultiScale(roi, 1.1, 4, minSize=ms):
            cys.append(y + ey + eh / 2.0)
    return float(np.median(cys)) if cys else (y + 0.38 * h)


def _clip(x0, y0, x1, y1, W, H):
    x0 = max(0, int(round(x0))); y0 = max(0, int(round(y0)))
    x1 = min(W, int(round(x1))); y1 = min(H, int(round(y1)))
    return x0, y0, max(0, x1 - x0), max(0, y1 - y0)


def locate_region(frames, cfg: Config, face=None) -> Region:
    """Return the glasses-region bbox in canonical (preprocessed) pixel coords."""
    H, W = frames.gray.shape[:2]
    if face is None:
        face = facedet.detect_face(frames.gray)

    # ---- WORN: face anchor + eye line ----
    if face is not None:
        eye_cy = _eye_line(frames.gray, face)
        fw, fh = face.w, face.h
        x0 = face.x - cfg.region_wpad * fw
        x1 = face.x + fw + cfg.region_wpad * fw
        y0 = eye_cy - cfg.region_up * fh
        y1 = eye_cy + cfg.region_down * fh
        x, y, w, h = _clip(x0, y0, x1, y1, W, H)
        if w > 20 and h > 20:
            return Region(x, y, w, h, "worn")

    # ---- HELD / STUDIO: dark-frame segmentation ----
    from pipeline import segment as _seg
    seg = _seg.segment(frames.gray_eq, cfg)
    if seg.valid:
        bx, by, bw, bh = seg.bbox
        pad_x = cfg.region_seg_pad * bw
        pad_y = cfg.region_seg_pad * bh
        x, y, w, h = _clip(bx - pad_x, by - pad_y, bx + bw + pad_x, by + bh + pad_y, W, H)
        return Region(x, y, w, h, "held")

    # ---- fallback: central 90% ----
    x, y, w, h = _clip(0.05 * W, 0.05 * H, 0.95 * W, 0.95 * H, W, H)
    return Region(x, y, w, h, "center")


def crop_region(frames, region: Region) -> np.ndarray:
    """BGR crop of the located region from the canonical colour image."""
    return frames.color[region.y:region.y + region.h, region.x:region.x + region.w]
